File: deal-mgmt/app/schemas/vdr.py
"""VDR (Virtual Data Room) Pydantic 스키마."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import VdrAccessAction, VdrClassificationStatus, VdrDocumentStatus, VdrFolderCategory

# ── 폴더 스키마 ───────────────────────────────────────────


class VdrFolderCreate(BaseModel):
    """VDR 폴더 생성 요청."""

    name: str = Field(..., min_length=1, max_length=255)
    category: VdrFolderCategory = VdrFolderCategory.CUSTOM
    parent_id: uuid.UUID | None = None
    description: str | None = Field(None, max_length=500)


class VdrFolderUpdate(BaseModel):
    """VDR 폴더 수정 요청 (부분 업데이트)."""

    name: str | None = Field(None, min_length=1, max_length=255)
    order_index: int | None = None
    description: str | None = Field(None, max_length=500)


class VdrFolderOut(BaseModel):
    """VDR 폴더 응답."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    transaction_id: uuid.UUID
    parent_id: uuid.UUID | None
    name: str
    category: VdrFolderCategory
    order_index: int
    is_required: bool
    description: str | None
    created_at: datetime
    updated_at: datetime


class VdrFolderTreeOut(VdrFolderOut):
    """트리 구조 응답 — 하위 폴더 + 문서 수 포함."""

    children: list[VdrFolderTreeOut] = []
    document_count: int = 0


class VdrInitRequest(BaseModel):
    """VDR 기본 폴더 구조 초기화 요청."""

    include_custom_folders: bool = False


# ── 문서 스키마 ───────────────────────────────────────────


class VdrDocumentOut(BaseModel):
    """VDR 문서(파일) 응답."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    transaction_id: uuid.UUID
    folder_id: uuid.UUID
    original_name: str
    file_size_bytes: int
    mime_type: str
    sha256_hash: str | None
    status: VdrDocumentStatus
    description: str | None
    uploaded_by_email: str | None
    classification_status: VdrClassificationStatus | None = None
    classification_score: int | None = None
    manual_review_needed: bool = False
    created_at: datetime
    updated_at: datetime


class VdrDocumentUpdate(BaseModel):
    """VDR 문서 수정 (설명, 폴더 이동)."""

    description: str | None = None
    folder_id: uuid.UUID | None = None


class VdrSummaryOut(BaseModel):
    """VDR 요약 통계."""

    total_folders: int = 0
    total_documents: int = 0
    total_size_bytes: int = 0
    initialized: bool = False


class VdrOverviewItem(BaseModel):
    """거래별 VDR 현황 — Overview 페이지용."""

    transaction_id: uuid.UUID
    transaction_name: str
    code_name: str
    phase: str
    status: str
    vdr_initialized: bool
    total_folders: int
    total_documents: int
    total_size_bytes: int
    last_upload_at: datetime | None


class VdrAutoUploadResult(BaseModel):
    """자동 라우팅 업로드 결과."""

    model_config = ConfigDict(from_attributes=True)

    document: VdrDocumentOut
    routed_folder: VdrFolderOut
    routed_category: VdrFolderCategory | None
    was_fallback: bool
    original_name_renamed: bool
    final_name: str


# ── Direct Upload 스키마 ──────────────────────────────────


class SuggestCategoryRequest(BaseModel):
    """파일명 기반 카테고리 추천 요청."""

    filename: str = Field(..., min_length=1, max_length=500)


class SuggestCategoryResponse(BaseModel):
    """파일명 기반 카테고리 추천 응답."""

    category: str | None = None
    folder_name: str | None = None


class FailedFileInfo(BaseModel):
    """Direct Upload 실패 파일 정보."""

    filename: str
    reason: str


class DirectUploadFileResult(BaseModel):
    """Direct Upload 개별 파일 분류 결과."""

    document: VdrDocumentOut
    routed_folder: VdrFolderOut
    routed_category: VdrFolderCategory | None
    classification_status: VdrClassificationStatus
    score: int
    was_fallback: bool


class DirectUploadBatchResult(BaseModel):
    """Direct Upload 다중 파일 배치 결과."""

    results: list[DirectUploadFileResult]
    pending_review_count: int
    total_uploaded: int
    failed_files: list[FailedFileInfo] = []


class ClassificationStatusOut(BaseModel):
    """2차 심사 상태 조회 응답."""

    document_id: uuid.UUID
    classification_status: VdrClassificationStatus
    routed_folder: VdrFolderOut | None = None
    routed_category: VdrFolderCategory | None = None
    manual_review_needed: bool


# ── 접근 로그 스키마 ──────────────────────────────────────


class VdrAccessLogOut(BaseModel):
    """VDR 접근 로그 응답."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    transaction_id: uuid.UUID
    document_id: uuid.UUID | None
    folder_id: uuid.UUID | None
    user_email: str
    user_id: str
    action: VdrAccessAction
    ip_address: str | None
    user_agent: str | None
    buyer_id: uuid.UUID | None
    created_at: datetime


class VdrAccessLogListResponse(BaseModel):
    """접근 로그 페이지네이션 응답."""

    items: list[VdrAccessLogOut]
    total: int


class BuyerActivitySummary(BaseModel):
    """매수자별 VDR 활동 요약."""

    buyer_id: uuid.UUID
    buyer_name: str
    unique_documents_accessed: int
    total_views: int
    total_downloads: int
    last_access_at: datetime | None

File: amic-platform/src/modules/ma/types/vdr.ts
// ── VDR 폴더 ─────────────────────────────────────────────

export type VdrFolderCategory =
  | "CORPORATE"
  | "FINANCIAL"
  | "LEGAL"
  | "TAX"
  | "HR"
  | "TECHNICAL"
  | "COMMERCIAL"
  | "REAL_ESTATE"
  | "ENVIRONMENT"
  | "IP"
  | "INSURANCE"
  | "MARKET_RESEARCH"
  | "CUSTOM";

/**
 * 서버는 중첩 트리(children 배열)로 응답하지만,
 * useVdrFolders 훅이 flattenVdrFolders로 평탄화하여 parent_id 참조 배열로 변환한다.
 * UI 컴포넌트는 평탄 배열을 받아 childrenByParentId Map으로 인덱싱해 사용한다.
 */
export interface VdrFolder {
  id: string;
  transaction_id: string;
  parent_id: string | null;
  name: string;
  category: VdrFolderCategory;
  order_index: number;
  is_required: boolean;
  description: string | null;
  created_at: string;
  updated_at: string;
  children: VdrFolder[];
  document_count: number;
}

export interface VdrFolderCreate {
  name: string;
  category?: VdrFolderCategory;
  parent_id?: string;
  description?: string;
}

export interface VdrFolderUpdate {
  name?: string;
  order_index?: number;
  description?: string;
}

// ── VDR 문서 ─────────────────────────────────────────────

export type VdrDocumentStatus = "ACTIVE" | "ARCHIVED" | "DELETED";

export type VdrClassificationStatus =
  | "DIRECT"
  | "PENDING_REVIEW"
  | "CLASSIFIED"
  | "MANUAL_REVIEW"
  | "MANUAL";

export interface VdrDocument {
  id: string;
  transaction_id: string;
  folder_id: string;
  original_name: string;
  file_size_bytes: number;
  mime_type: string;
  sha256_hash: string | null;
  status: VdrDocumentStatus;
  description: string | null;
  uploaded_by_email: string | null;
  classification_status: VdrClassificationStatus | null;
  classification_score: number | null;
  manual_review_needed: boolean;
  created_at: string;
  updated_at: string;
}

export interface VdrDocumentUpdate {
  description?: string;
  folder_id?: string;
}

export type VdrWorkstream = "LDD" | "FDD" | "VALUATION" | "COMMON";
export type VdrRoutingQueueStatus = "open" | "reviewed" | "all";

export interface VdrRoutingOverride {
  id: string;
  transaction_id: string;
  vdr_document_id: string;
  primary_workstream: VdrWorkstream;
  workstream_tags: VdrWorkstream[];
  override_note: string | null;
  reviewed_by_email: string | null;
  reviewed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface VdrRoutingOverrideUpsert {
  primary_workstream: VdrWorkstream;
  workstream_tags?: VdrWorkstream[];
  override_note?: string | null;
}

export interface VdrRoutingDecision {
  primary_workstream: VdrWorkstream;
  workstream_tags: VdrWorkstream[];
  confidence: number;
  requires_manual_review: boolean;
  reasons: string[];
  is_override: boolean;
  override_note: string | null;
  reviewed_by_email: string | null;
  reviewed_at: string | null;
}

export interface VdrRoutingQueueItem {
  document: VdrDocument;
  folder_name: string;
  folder_category: VdrFolderCategory;
  routing_status: "OPEN_REVIEW" | "OVERRIDDEN" | "AUTO_ROUTED";
  auto_route: VdrRoutingDecision;
  effective_route: VdrRoutingDecision;
}

export interface VdrRoutingQueueSummary {
  total_documents: number;
  returned_documents: number;
  open_documents: number;
  reviewed_documents: number;
  auto_routed_documents: number;
  by_effective_workstream: Partial<Record<VdrWorkstream, number>>;
}

export interface VdrRoutingQueueResponse {
  summary: VdrRoutingQueueSummary;
  items: VdrRoutingQueueItem[];
}

// ── Direct Upload ────────────────────────────────────────

export interface DirectUploadFileResult {
  document: VdrDocument;
  routed_folder: VdrFolder;
  routed_category: VdrFolderCategory | null;
  classification_status: VdrClassificationStatus;
  score: number;
  was_fallback: boolean;
}

export interface FailedFileInfo {
  filename: string;
  reason: string;
}

export interface DirectUploadBatchResult {
  results: DirectUploadFileResult[];
  pending_review_count: number;
  total_uploaded: number;
  failed_files: FailedFileInfo[];
}

export interface ClassificationStatusItem {
  document_id: string;
  classification_status: VdrClassificationStatus;
  routed_folder: VdrFolder | null;
  routed_category: VdrFolderCategory | null;
  manual_review_needed: boolean;
}

// ── VDR 요약 ─────────────────────────────────────────────

export interface VdrSummary {
  total_folders: number;
  total_documents: number;
  total_size_bytes: number;
  initialized: boolean;
}

// ── 상수 ─────────────────────────────────────────────────

export const VDR_CATEGORY_LABELS: Record<VdrFolderCategory, string> = {
  CORPORATE: "기업 일반",
  FINANCIAL: "재무 자료",
  LEGAL: "법률 자료",
  TAX: "세무 자료",
  HR: "인사/노무",
  TECHNICAL: "기술/IT",
  COMMERCIAL: "영업/마케팅",
  REAL_ESTATE: "부동산/자산",
  ENVIRONMENT: "환경",
  IP: "지식재산권",
  INSURANCE: "보험",
  MARKET_RESEARCH: "시장자료",
  CUSTOM: "사용자 정의",
};

export const VDR_WORKSTREAM_LABELS: Record<VdrWorkstream, string> = {
  LDD: "LDD",
  FDD: "FDD",
  VALUATION: "Valuation",
  COMMON: "Common",
};

export const MIME_TYPE_LABELS: Record<string, string> = {
  "application/pdf": "PDF",
  "application/vnd.openxmlformats-officedocument.wordprocessingml.document":
    "Word",
  "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet": "Excel",
  "application/vnd.openxmlformats-officedocument.presentationml.presentation":
    "PPT",
  "image/png": "PNG",
  "image/jpeg": "JPEG",
  "text/csv": "CSV",
  "text/plain": "TXT",
  "application/zip": "ZIP",
};

/** 2차 심사 상태 폴링 간격 (ms) */
export const CLASSIFICATION_POLL_INTERVAL_MS = 5000;

// ── VDR 접근 추적 ────────────────────────────────────────
export type VdrAccessAction = "VIEW" | "DOWNLOAD" | "UPLOAD";

export interface VdrAccessLog {
  id: string;
  transaction_id: string;
  document_id: string | null;
  folder_id: string | null;
  user_email: string;
  user_id: string;
  action: VdrAccessAction;
  ip_address: string | null;
  user_agent: string | null;
  buyer_id: string | null;
  created_at: string;
}

export interface VdrAccessLogListResponse {
  items: VdrAccessLog[];
  total: number;
}

export interface BuyerActivitySummary {
  buyer_id: string;
  buyer_name: string;
  unique_documents_accessed: number;
  total_views: number;
  total_downloads: number;
  last_access_at: string | null;
}

/** VDR 파일 업로드 제약 조건 — 백엔드와 동기화 필수 */
export const VDR_CONSTRAINTS = {
  /** 최대 파일 크기 (bytes): 100MB */
  MAX_FILE_SIZE: 100 * 1024 * 1024,
  /** 최대 파일 크기 (표시용) */
  MAX_FILE_SIZE_LABEL: "100MB",
  /** 허용 MIME 타입 — deal-mgmt/app/routers/vdr.py _ALLOWED_MIME_TYPES 동기화 */
  ALLOWED_MIME_TYPES: new Set([
    "application/pdf",
    "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    "application/vnd.openxmlformats-officedocument.presentationml.presentation",
    "application/msword",
    "application/vnd.ms-excel",
    "application/vnd.ms-powerpoint",
    "image/png",
    "image/jpeg",
    "text/plain",
    "text/csv",
    "application/zip",
    // M&A 실사 특화 포맷 (BE와 동기화)
    "application/haansofthwp", // .hwp
    "application/x-hwp", // .hwp 변형
    "image/vnd.dwg", // .dwg
    "application/dxf", // .dxf
    "application/x-tar", // .tar
    "application/gzip", // .gz
    "application/json", // .json
    "application/vnd.google-earth.kml+xml", // .kml
    "application/x-iwork-keynote-sffkey", // .key
    "message/rfc822", // .eml
    "application/vnd.ms-outlook", // .msg
  ]),
  /** 허용 파일 확장자 (input accept 속성용) — BE와 동기화 */
  ACCEPT_EXTENSIONS:
    ".pdf,.docx,.xlsx,.pptx,.doc,.xls,.ppt,.png,.jpg,.jpeg,.txt,.csv,.zip,.hwp,.dwg,.dxf,.tar,.gz,.json,.kml,.key,.eml,.msg",
} as const;
